experimental/dds/tree2/src/feature-libraries/chunked-forest/codec/chunkDecoding.ts

```diff
@@ -0,0 +1,257 @@
+>({
+	a(shape: EncodedNestedArray, cache): ChunkDecoder {
+		return new NestedArrayDecoder(shape);
+	},
+	b(shape: EncodedInlineArray, cache): ChunkDecoder {
+		return new InlineArrayDecoder(shape);
+	},
+	c(shape: EncodedTreeShape, cache): ChunkDecoder {
+		return new TreeDecoder(shape, cache);
+	},
+	d(shape: EncodedAnyShape): ChunkDecoder {
+		return AnyDecoder.instance;
+	},
+});
+
+export function readValue(stream: StreamCursor, shape: EncodedValueShape): Value {
+	if (shape === undefined) {
+		return readStreamBoolean(stream) ? readStream(stream) : undefined;
+	} else {
+		if (shape === true) {
+			return readStream(stream);
+		} else if (shape === false) {
+			return undefined;
+		} else if (Array.isArray(shape)) {
+			assert(shape.length === 1, "expected a single constant for value");
+			return shape[0] as Value;
+		} else {
+			// EncodedCounter case:
+			unreachableCase(shape, "decoding values as deltas is not yet supported");
+		}
+	}
+}
+
+export function deaggregateChunks(chunk: TreeChunk): TreeChunk[] {
+	if (chunk === emptyChunk) {
+		return [];
+	}
+	// TODO: when handling of SequenceChunks has better performance (for example in cursors),
+	// consider keeping SequenceChunks here if they are longer than some threshold.
+	if (chunk instanceof SequenceChunk) {
+		// Could return [] here, however the logic in this file is designed to never produce an empty SequenceChunk, so its better to throw an error here to detect bugs.
+		assert(chunk.subChunks.length > 0, "Unexpected empty sequence");
+		// Logic in this file is designed to never produce an unneeded (single item) SequenceChunks, so its better to throw an error here to detect bugs.
+		assert(chunk.subChunks.length > 1, "Unexpected single item sequence");
+
+		for (const sub of chunk.subChunks) {
+			// The logic in this file is designed to never produce an nested SequenceChunks or emptyChunk, so its better to throw an error here to detect bugs.
+			assert(!(sub instanceof SequenceChunk), "unexpected nested sequence");
+			assert(sub !== emptyChunk, "unexpected empty chunk");
+
+			sub.referenceAdded();
+		}
+
+		chunk.referenceRemoved();
+		return chunk.subChunks;
+	} else {
+		return [chunk];
+	}
+}
+
+export function aggregateChunks(input: TreeChunk[]): TreeChunk {
+	const chunks = input.flatMap(deaggregateChunks);
+	switch (chunks.length) {
+		case 0:
+			return emptyChunk;
+		case 1:
+			return chunks[0];
+		default:
+			return new SequenceChunk(chunks);
+	}
+}
+
+export class NestedArrayDecoder implements ChunkDecoder {
+	public constructor(private readonly shape: EncodedNestedArray) {}
+	public decode(decoders: readonly ChunkDecoder[], stream: StreamCursor): TreeChunk {
+		const decoder = decoders[this.shape];
+
+		// TODO: uniform chunk fast path
+		const chunks: TreeChunk[] = [];
+
+		const data = readStream(stream);
+		if (typeof data === "number") {
+			// This case means that the array contained 0 sized items, and was thus encoded as the length instead of the array.
```

Comment:
nit: while correct, this comment could be easily mistaken for "0-sized items".
```suggestion
			// This case means that the array contained only 0-sized items, and was thus encoded as the length of the array.
```

(or equivalent)